### ISSUE TITLE
S2699: should be skipped on tests with TestCaseSourceAttribute and return values

### ISSUE THREAD

**Issue by alexandersorokin** [SonarSource/sonar-dotnet#3751]
### Description

Rule ID: S2699

It does not respect `NUnit` `TestCaseSourceAttribute` for methods with return value.

### Repro steps
1. Reference NUnit 3;
2. Run following code:
```
[TestFixture]
internal class Equals_Should
{
  private static readonly TestCaseData[] Cases =
  {
    new TestCaseData("foo", "bar").Returns(false),
    new TestCaseData(100, 100).Returns(true),
  };

  [Test]
  [TestCaseSource(nameof(Cases))]
  public bool Test(object obj1, object obj2)
  {
    return obj1.Equals(obj2);
  }
}
```

### Expected behavior

No warnings as return value is checked by NUnit.

### Actual behavior

```
Warning	S2699	Add at least one assertion to this test case
```

### Known workarounds

No workarounds was found.

### Related information

* C# 8
* Visual Studio 2019 Pro. Version 16.7.6
* donetcore3.1
* Windows 10. Version 20H2.

**Comment by alexandersorokin** [comment]
So, I think that the `S2699` rule should run on test methods with `TestCaseSourceAttribute` only if the test method returns `void` or `Task`.

**Comment by costin-zaharia-sonarsource** [comment]
Hi @alexandersorokin. Thanks for your feedback. I can confirm the false positive.

Would you be interested to write a fix for it?

Thanks,
Costin

**Comment by alexandersorokin** [comment]
Yes, I can fix it. But firstly we should discuss how it should be fixed.

I think that the right way to fix the bug is to split the rule `S2699` into two rules.

The first rule should ensure that a test method has at least one assertion or has a return value. So it is going to be the same as the currently implemented rule `S2699` apart from two things. Firstly, it should not make use of `TestCaseAttribute`. Secondly, it should not show the warning if a test method has a return value.
I can do it by replacing the check `methodSymbol.HasAssertionInAttribute()` of `TestCaseAttribute` at line 70 in [TestMethodShouldContainAssertion.cs](https://github.com/SonarSource/sonar-dotnet/blob/8f7235325e97b80c176a7ff170facdbc7a229832/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldContainAssertion.cs#L70) with check of a value of `methodSymbol.ReturnType` property. If the type is `void` or it is a non-generic version of `Task`, the rule evaluator should continue to try to find an assertion in the test method. If it is not `void` and it is not a non-generic version of `Task`, then there are no requirements for that test method to contain an assertion.

The second rule should ensure the following: if a test method has a return value (it is not `void` and it is not a non-generic version of `Task`) then the test method should have `TestCaseSourceAttribute` or `TestCaseAttribute` with assigned `ExpectedResult` property. But currently, I am not allowed to append such a new rule because I should request it at the community forum as stated in [README.md](https://github.com/SonarSource/sonar-dotnet/blob/master/README.md#custom-rules) of this repository. And I am not eager to write something there.
If you approve the addition of that rule then I think I can implement it.

**Comment by costin-zaharia-sonarsource** [comment]
Hi, we discussed this briefly inside the team and think that it would be better to have one single rule for this. Using `TestCaseSourceAttribute` or `TestCaseAttribute` with `ExpectedResult` is still an assertion, just a different way to write it. 

I think `HasAssertionInAttribute` can be updated to check for `TestCaseSourceAttribute` too and the method return value. 

What do you think about this approach?

**Comment by alexandersorokin** [comment]
If I understand you correctly and you mean checking in `HasAssertionInAttribute` a presence of `TestCase` or a presence of a return value or a presence of `TestCaseSource`, then your variant would be much easier to implement but it would give false negatives in the two following cases:
```
    [TestFixture]
    internal class Something_Should
    {
        private static readonly TestCaseData[] Cases =
        {
            new TestCaseData("foo"),
            new TestCaseData(100),
        };

        [Test]
        [TestCaseSource(nameof(Cases))]
        public void Test1(object obj)
        {
            Console.WriteLine(obj);
        }

        [Test]
        [TestCase(ExpectedResult = true)]
        public void Test2()
        {
            Console.WriteLine();
        }
    }
}
```
Moreover, `Test2` is false negative currently.

I think that the assertion of a test method is a declaration of the return value. I am sure that any attributes of the test method are not assertions.

Also, I think that checking both (an attribute and return value) is redundant.

Actually replacing `methodSymbol.HasAssertionInAttribute()` with a check of the existence of the return value of a test method is enough to fix this bug. It is an easy task. But if we do it without implementing the second part of my proposal then there would be nothing to ensure that the test method has assigned `ExpectedResult` property in `TestCaseAttibute`. So it would introduce a regression.

--------------

If you mean checking an existence in `HasAssertionInAttribute` either `TestCase` and a return value or `TestCaseSource` and a return value then it would solve all issues written above. But another issue would remain:
```
    [TestFixture]
    internal class Something_Should
    {
        [Test]
        [TestCase(ExpectedResult = true)] // ExpectedResult is an error here.
        public void Test2()
        {
            Assert.IsTrue(true);
        }
    }
}
```
To fix it another rule is needed.

**Comment by alexandersorokin** [comment]
@costin-zaharia-sonarsource I also want to add that I do not mind implementing your variant. I just think that my variant covers more cases.

**Comment by costin-zaharia-sonarsource** [comment]
Hi @alexandersorokin, 

since time is limited we often need to chose between FPs and FNs and choosing FNs over FPs it's usually a trade off we are happy with as long as the rule as a good proportion of true positives.

Having one or multiple rules is a business decision and we would like to keep it as it is for now. Your suggestion with splitting the rule is one possible technical solution for fixing part of the FNs but I believe we could also rewrite the existing rule to cover the cases you described. We are not limited in changing only a method implementation :)

I think that the test cases you provided are valuable and I also think that there are more which we could consider since the test cases can be defined a partial class, in another type or can even be another type. They could also be defined in an assembly for which we don't have sources.

I suggest that we can start with baby steps, documenting the cases we know that aren't covered and fix the false positive.

**Comment by alexandersorokin** [comment]
WIP

**Comment by alexandersorokin** [comment]
@costin-zaharia-sonarsource PR: https://github.com/SonarSource/sonar-dotnet/pull/3799

**Comment by alexandersorokin** [comment]
@costin-zaharia-sonarsource : are there any updates on PR?

**Comment by alexandersorokin** [comment]
When are changes expected to be published at nuget?

**Comment by costin-zaharia-sonarsource** [comment]
Hi @alexandersorokin, thanks a lot for you contribution.

Our plan is to finish the current sprint by December 18th and the release should follow shortly after (if not that Friday it will most probably done the following Monday)
